{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ff5703a2",
   "metadata": {},
   "source": [
    "## Run differr\n",
    "\n",
    "Do pairwise comparisons between the strains."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "64a4fbbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>SampleName</th>\n",
       "      <th>Species</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Name</th>\n",
       "      <th>AssemblyID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13212615</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13212616</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13212618</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13212623</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13212624</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>SRR16608236</td>\n",
       "      <td>Bile1</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>RIMD</td>\n",
       "      <td>Vibrioparahaemolyticus-RIMD_2210633</td>\n",
       "      <td>GCF_000196095.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>SRR16608237</td>\n",
       "      <td>LB2</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>RIMD</td>\n",
       "      <td>Vibrioparahaemolyticus-RIMD_2210633</td>\n",
       "      <td>GCF_000196095.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>SRR16608238</td>\n",
       "      <td>LB1</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>RIMD</td>\n",
       "      <td>Vibrioparahaemolyticus-RIMD_2210633</td>\n",
       "      <td>GCF_000196095.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>SRR18432816</td>\n",
       "      <td>GSM5965598</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>USA300</td>\n",
       "      <td>Staphylococcusaureus-USA300</td>\n",
       "      <td>GCF_022226995.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SRR18432817</td>\n",
       "      <td>GSM5965599</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>USA300</td>\n",
       "      <td>Staphylococcusaureus-USA300</td>\n",
       "      <td>GCF_022226995.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Run           SampleName         Species  Strain  \\\n",
       "0   SRR13212615  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "1   SRR13212616  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "2   SRR13212618  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "3   SRR13212623  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "4   SRR13212624  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "..          ...                  ...             ...     ...   \n",
       "79  SRR16608236                Bile1          Vibrio    RIMD   \n",
       "80  SRR16608237                  LB2          Vibrio    RIMD   \n",
       "81  SRR16608238                  LB1          Vibrio    RIMD   \n",
       "82  SRR18432816           GSM5965598  Staphylococcus  USA300   \n",
       "83  SRR18432817           GSM5965599  Staphylococcus  USA300   \n",
       "\n",
       "                                   Name       AssemblyID  \n",
       "0             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "1             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "2             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "3             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "4             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "..                                  ...              ...  \n",
       "79  Vibrioparahaemolyticus-RIMD_2210633  GCF_000196095.1  \n",
       "80  Vibrioparahaemolyticus-RIMD_2210633  GCF_000196095.1  \n",
       "81  Vibrioparahaemolyticus-RIMD_2210633  GCF_000196095.1  \n",
       "82          Staphylococcusaureus-USA300  GCF_022226995.1  \n",
       "83          Staphylococcusaureus-USA300  GCF_022226995.1  \n",
       "\n",
       "[84 rows x 6 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get pairwise comparisons\n",
    "import pandas as pd\n",
    "run_to_ass = pd.read_csv(f'../output/run_to_assembly.csv')\n",
    "run_to_ass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28902694",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>DATASTORE provider</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue</th>\n",
       "      <th>geographic_location_(country_and/or_sea)</th>\n",
       "      <th>host_health_state</th>\n",
       "      <th>host_scientific_name</th>\n",
       "      <th>Host_disease</th>\n",
       "      <th>BIOMATERIAL_PROVIDER</th>\n",
       "      <th>media</th>\n",
       "      <th>Note</th>\n",
       "      <th>treatment</th>\n",
       "      <th>type-material</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13212615</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>1096</td>\n",
       "      <td>1522106869</td>\n",
       "      <td>SAMN17023022</td>\n",
       "      <td>1454555653</td>\n",
       "      <td>NEW ENGLAND BIOLABS, INC</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>s3,ncbi,gs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13212616</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>979</td>\n",
       "      <td>701284083</td>\n",
       "      <td>SAMN17023022</td>\n",
       "      <td>674896977</td>\n",
       "      <td>NEW ENGLAND BIOLABS, INC</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>ncbi,s3,gs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13212618</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>717</td>\n",
       "      <td>137491381</td>\n",
       "      <td>SAMN17023022</td>\n",
       "      <td>135168447</td>\n",
       "      <td>NEW ENGLAND BIOLABS, INC</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13212623</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>1096</td>\n",
       "      <td>1522106774</td>\n",
       "      <td>SAMN17023022</td>\n",
       "      <td>1454222338</td>\n",
       "      <td>NEW ENGLAND BIOLABS, INC</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>ncbi,gs,s3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13212624</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>979</td>\n",
       "      <td>701284088</td>\n",
       "      <td>SAMN17023022</td>\n",
       "      <td>674731029</td>\n",
       "      <td>NEW ENGLAND BIOLABS, INC</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>ncbi,s3,gs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>SRR16608236</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>822</td>\n",
       "      <td>851534610</td>\n",
       "      <td>SAMN22630217</td>\n",
       "      <td>626320528</td>\n",
       "      <td>OSAKA UNIVERSTIY</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>ncbi,s3,gs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>SRR16608237</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>879</td>\n",
       "      <td>1002194957</td>\n",
       "      <td>SAMN22630216</td>\n",
       "      <td>732648027</td>\n",
       "      <td>OSAKA UNIVERSTIY</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>ncbi,gs,s3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>SRR16608238</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>873</td>\n",
       "      <td>1519514091</td>\n",
       "      <td>SAMN22630215</td>\n",
       "      <td>1107523038</td>\n",
       "      <td>OSAKA UNIVERSTIY</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>ncbi,s3,gs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>SRR18432816</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>380</td>\n",
       "      <td>840405187</td>\n",
       "      <td>SAMN26871989</td>\n",
       "      <td>709489358</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>ncbi,gs,s3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TSB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sample harvested at OD600 = 3 after growth in TSB</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SRR18432817</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>359</td>\n",
       "      <td>1024530718</td>\n",
       "      <td>SAMN26871990</td>\n",
       "      <td>898634960</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>gs,ncbi,s3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human serum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sample harvested at OD600 = 3 after growth in TSB</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Run Assay Type  AvgSpotLen       Bases     BioSample       Bytes  \\\n",
       "0   SRR13212615    RNA-Seq        1096  1522106869  SAMN17023022  1454555653   \n",
       "1   SRR13212616    RNA-Seq         979   701284083  SAMN17023022   674896977   \n",
       "2   SRR13212618    RNA-Seq         717   137491381  SAMN17023022   135168447   \n",
       "3   SRR13212623    RNA-Seq        1096  1522106774  SAMN17023022  1454222338   \n",
       "4   SRR13212624    RNA-Seq         979   701284088  SAMN17023022   674731029   \n",
       "..          ...        ...         ...         ...           ...         ...   \n",
       "79  SRR16608236    RNA-Seq         822   851534610  SAMN22630217   626320528   \n",
       "80  SRR16608237    RNA-Seq         879  1002194957  SAMN22630216   732648027   \n",
       "81  SRR16608238    RNA-Seq         873  1519514091  SAMN22630215  1107523038   \n",
       "82  SRR18432816    RNA-Seq         380   840405187  SAMN26871989   709489358   \n",
       "83  SRR18432817    RNA-Seq         359  1024530718  SAMN26871990   898634960   \n",
       "\n",
       "                 Center Name Consent DATASTORE filetype DATASTORE provider  \\\n",
       "0   NEW ENGLAND BIOLABS, INC  public          fastq,sra         s3,ncbi,gs   \n",
       "1   NEW ENGLAND BIOLABS, INC  public          sra,fastq         ncbi,s3,gs   \n",
       "2   NEW ENGLAND BIOLABS, INC  public          fastq,sra         gs,ncbi,s3   \n",
       "3   NEW ENGLAND BIOLABS, INC  public          fastq,sra         ncbi,gs,s3   \n",
       "4   NEW ENGLAND BIOLABS, INC  public          fastq,sra         ncbi,s3,gs   \n",
       "..                       ...     ...                ...                ...   \n",
       "79          OSAKA UNIVERSTIY  public          fastq,sra         ncbi,s3,gs   \n",
       "80          OSAKA UNIVERSTIY  public          fastq,sra         ncbi,gs,s3   \n",
       "81          OSAKA UNIVERSTIY  public          fastq,sra         ncbi,s3,gs   \n",
       "82                       GEO  public          fastq,sra         ncbi,gs,s3   \n",
       "83                       GEO  public          sra,fastq         gs,ncbi,s3   \n",
       "\n",
       "    ... tissue geographic_location_(country_and/or_sea) host_health_state  \\\n",
       "0   ...    NaN                                      NaN               NaN   \n",
       "1   ...    NaN                                      NaN               NaN   \n",
       "2   ...    NaN                                      NaN               NaN   \n",
       "3   ...    NaN                                      NaN               NaN   \n",
       "4   ...    NaN                                      NaN               NaN   \n",
       "..  ...    ...                                      ...               ...   \n",
       "79  ...    NaN                                      NaN               NaN   \n",
       "80  ...    NaN                                      NaN               NaN   \n",
       "81  ...    NaN                                      NaN               NaN   \n",
       "82  ...    NaN                                      NaN               NaN   \n",
       "83  ...    NaN                                      NaN               NaN   \n",
       "\n",
       "   host_scientific_name Host_disease BIOMATERIAL_PROVIDER        media Note  \\\n",
       "0                   NaN          NaN                  NaN          NaN  NaN   \n",
       "1                   NaN          NaN                  NaN          NaN  NaN   \n",
       "2                   NaN          NaN                  NaN          NaN  NaN   \n",
       "3                   NaN          NaN                  NaN          NaN  NaN   \n",
       "4                   NaN          NaN                  NaN          NaN  NaN   \n",
       "..                  ...          ...                  ...          ...  ...   \n",
       "79                  NaN          NaN                  NaN          NaN  NaN   \n",
       "80                  NaN          NaN                  NaN          NaN  NaN   \n",
       "81                  NaN          NaN                  NaN          NaN  NaN   \n",
       "82                  NaN          NaN                  NaN          TSB  NaN   \n",
       "83                  NaN          NaN                  NaN  Human serum  NaN   \n",
       "\n",
       "                                            treatment type-material  \n",
       "0                                                 NaN           NaN  \n",
       "1                                                 NaN           NaN  \n",
       "2                                                 NaN           NaN  \n",
       "3                                                 NaN           NaN  \n",
       "4                                                 NaN           NaN  \n",
       "..                                                ...           ...  \n",
       "79                                                NaN           NaN  \n",
       "80                                                NaN           NaN  \n",
       "81                                                NaN           NaN  \n",
       "82  Sample harvested at OD600 = 3 after growth in TSB           NaN  \n",
       "83  Sample harvested at OD600 = 3 after growth in TSB           NaN  \n",
       "\n",
       "[73 rows x 67 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "data = '../data_refs/'\n",
    "meta = pd.read_csv(f'{data}SraRunTable_RNAseq_bacteria_oxfordnanopore.csv')\n",
    "meta\n",
    "\n",
    "run_to_org = dict(zip(meta.Run, meta.Organism)) #meta[meta['Run'] == 'SRR16009459']['Organism']\n",
    "run_to_strain = dict(zip(meta.Run, meta.strain))\n",
    "\n",
    "mapping_dir = '../output/mapping/'\n",
    "runs = meta['Run'].values\n",
    "mapping = []\n",
    "columns = ['run', 'Organism', 'strain', 'reference_type', 'total', 'secondary', 'supplementary', 'duplicates', 'mapped', 'percent']\n",
    "for run in runs:\n",
    "    with open(f'{mapping_dir}{run}_minimap2_transcriptomic-quality.txt', 'r') as f:\n",
    "        row = [run, run_to_org[run], run_to_strain[run], 'transcriptome']\n",
    "        c = 0\n",
    "        for line in f:\n",
    "            line = line.split(' ')\n",
    "            if c < 4:\n",
    "                row.append(int(line[0]))\n",
    "            elif c == 4:\n",
    "                mapped = int(line[0])\n",
    "                perc = float(line[4].split(\"(\")[1].split('%')[0])\n",
    "                row.append(mapped)\n",
    "                row.append(perc)\n",
    "                break\n",
    "            c += 1\n",
    "    mapping.append(row)\n",
    "    with open(f'{mapping_dir}{run}_minimap2_genome-quality.txt', 'r') as f:\n",
    "        row = [run, run_to_org[run], run_to_strain[run], 'genome']\n",
    "        c = 0\n",
    "        for line in f:\n",
    "            line = line.split(' ')\n",
    "            if c < 4:\n",
    "                row.append(int(line[0]))\n",
    "            elif c == 4:\n",
    "                mapped = int(line[0])\n",
    "                perc = float(line[4].split(\"(\")[1].split('%')[0])\n",
    "                row.append(mapped)\n",
    "                row.append(perc)\n",
    "                break\n",
    "            c += 1\n",
    "    mapping.append(row)\n",
    "    \n",
    "mapping_df = pd.DataFrame(data=mapping, columns=columns)\n",
    "runs = meta[meta['Run'].isin(mapping_df[mapping_df['percent'] > 70]['run'].values)]\n",
    "runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8eb0521e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Listeria monocytogenes                  58\n",
       "Escherichia coli                        28\n",
       "Streptococcus mutans                    24\n",
       "Xylella fastidiosa subsp. fastidiosa    22\n",
       "Pseudomonas aeruginosa UCBPP-PA14       12\n",
       "Klebsiella pneumoniae                    8\n",
       "Vibrio parahaemolyticus                  8\n",
       "Pandoraea fibrosis                       4\n",
       "Staphylococcus aureus                    4\n",
       "Name: Organism, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_df['Organism'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "169782e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Listeriamonocytogenes-ref              29\n",
       "Escherichiacoli-K_12                   12\n",
       "Streptococcusmutans-B04Sm5             12\n",
       "Xylellafastidiosa-stags_leap           11\n",
       "Pseudomonasaeruginosa-UCBPP_PA14        6\n",
       "Vibrioparahaemolyticus-RIMD_2210633     4\n",
       "Escherichiacoli-E2348_69                2\n",
       "Staphylococcusaureus-USA300             2\n",
       "Klebsiellapneumoniae-16_GR_13           1\n",
       "Klebsiellapneumoniae-20_GR_12           1\n",
       "Klebsiellapneumoniae-1_GR_13            1\n",
       "Klebsiellapneumoniae-2_GR_12            1\n",
       "Pandoraeafibrosis-6399                  1\n",
       "Pandoraeafibrosis-7641                  1\n",
       "Name: Name, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_to_ass['Name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69feb7ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000005845.2']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_to_ass = run_to_ass[run_to_ass['Run'].isin(runs['Run'].values)]\n",
    "# Group by the assembly id\n",
    "assemblies = run_to_ass[run_to_ass['Name'] == 'Escherichiacoli-K_12']\n",
    "assemblies = list(set(assemblies['AssemblyID'].values))\n",
    "assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f40ffd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each of the runs that maps to these do a pairwise differr comp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8e52facb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: differr [OPTIONS]\r\n",
      "\r\n",
      "  A script for detecting differential error rates in aligned Nanopore data\r\n",
      "\r\n",
      "Options:\r\n",
      "  -a, --cond-A-bams TEXT          BAM files for low modification sample. To\r\n",
      "                                  include multiple BAMs you need to use the -a\r\n",
      "                                  flag multiple times.  [required]\r\n",
      "  -b, --cond-B-bams TEXT          BAM files for wild type/complemented line\r\n",
      "                                  with normal modification levels  [required]\r\n",
      "  -r, --reference-fasta TEXT      [required]\r\n",
      "  -o, --output-bed TEXT           [required]\r\n",
      "  -c, --raw-counts-hdf TEXT\r\n",
      "  -f, --fdr-threshold FLOAT\r\n",
      "  -p, --processes INTEGER\r\n",
      "  -m, --max-depth INTEGER         Maximum depth for pysam pileup, default is a\r\n",
      "                                  very large number (basically samples all\r\n",
      "                                  reads)\r\n",
      "  --normalise / --no-normalise    Whether to normalise counts by sample depth\r\n",
      "                                  (counts per million) before performing\r\n",
      "                                  stats.\r\n",
      "  --median-expr-threshold INTEGER\r\n",
      "                                  The minimum value for the median number of\r\n",
      "                                  reads per condition after normalisation.\r\n",
      "                                  Default is 10.\r\n",
      "  --min-expr-threshold INTEGER    The minimum number of reads that ALL\r\n",
      "                                  replicates should have after normalisation.\r\n",
      "                                  Default is 1.\r\n",
      "  -q, --query TEXT\r\n",
      "  --help                          Show this message and exit.\r\n"
     ]
    }
   ],
   "source": [
    "! differr --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9801931b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>SampleName</th>\n",
       "      <th>Species</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Name</th>\n",
       "      <th>AssemblyID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13212615</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13212616</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13212618</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13212623</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13212624</td>\n",
       "      <td>01_Listeria_RNA_seq</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Listeriamonocytogenes-ref</td>\n",
       "      <td>GCF_000196035.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>SRR16608236</td>\n",
       "      <td>Bile1</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>RIMD</td>\n",
       "      <td>Vibrioparahaemolyticus-RIMD_2210633</td>\n",
       "      <td>GCF_000196095.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>SRR16608237</td>\n",
       "      <td>LB2</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>RIMD</td>\n",
       "      <td>Vibrioparahaemolyticus-RIMD_2210633</td>\n",
       "      <td>GCF_000196095.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>SRR16608238</td>\n",
       "      <td>LB1</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>RIMD</td>\n",
       "      <td>Vibrioparahaemolyticus-RIMD_2210633</td>\n",
       "      <td>GCF_000196095.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>SRR18432816</td>\n",
       "      <td>GSM5965598</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>USA300</td>\n",
       "      <td>Staphylococcusaureus-USA300</td>\n",
       "      <td>GCF_022226995.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SRR18432817</td>\n",
       "      <td>GSM5965599</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>USA300</td>\n",
       "      <td>Staphylococcusaureus-USA300</td>\n",
       "      <td>GCF_022226995.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Run           SampleName         Species  Strain  \\\n",
       "0   SRR13212615  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "1   SRR13212616  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "2   SRR13212618  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "3   SRR13212623  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "4   SRR13212624  01_Listeria_RNA_seq        Listeria     NaN   \n",
       "..          ...                  ...             ...     ...   \n",
       "79  SRR16608236                Bile1          Vibrio    RIMD   \n",
       "80  SRR16608237                  LB2          Vibrio    RIMD   \n",
       "81  SRR16608238                  LB1          Vibrio    RIMD   \n",
       "82  SRR18432816           GSM5965598  Staphylococcus  USA300   \n",
       "83  SRR18432817           GSM5965599  Staphylococcus  USA300   \n",
       "\n",
       "                                   Name       AssemblyID  \n",
       "0             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "1             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "2             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "3             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "4             Listeriamonocytogenes-ref  GCF_000196035.1  \n",
       "..                                  ...              ...  \n",
       "79  Vibrioparahaemolyticus-RIMD_2210633  GCF_000196095.1  \n",
       "80  Vibrioparahaemolyticus-RIMD_2210633  GCF_000196095.1  \n",
       "81  Vibrioparahaemolyticus-RIMD_2210633  GCF_000196095.1  \n",
       "82          Staphylococcusaureus-USA300  GCF_022226995.1  \n",
       "83          Staphylococcusaureus-USA300  GCF_022226995.1  \n",
       "\n",
       "[73 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_to_ass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2c4146c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "genome_dir = '../genomes/'\n",
    "run_to_gff = {}\n",
    "run_to_fna = {}\n",
    "ass_to_gff = {}\n",
    "ass_to_ref = {}\n",
    "run_to_assembly = dict(zip(run_to_ass['Run'], run_to_ass['AssemblyID'] ))\n",
    "files = os.listdir(genome_dir)\n",
    "for run in run_to_assembly:\n",
    "    assembly = run_to_assembly[run]\n",
    "    for f in files:\n",
    "        if assembly in f and 'fna' in f and 'fai' not in f:\n",
    "            run_to_fna[run] = f'{genome_dir}{f}'\n",
    "            ass_to_ref[assembly] = f.replace('.fna', '_transcripts.fasta')\n",
    "        elif assembly in f and 'gff' in f:\n",
    "            run_to_gff[run] = f'{genome_dir}{f}'\n",
    "            ass_to_gff[assembly] = f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1dc4adf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('SRR12090426', 'SRR12090427')\n",
      "('SRR12090426', 'SRR12090428')\n",
      "('SRR12090426', 'SRR12090429')\n",
      "('SRR12090426', 'SRR12090430')\n",
      "('SRR12090426', 'SRR12090431')\n",
      "('SRR12090426', 'SRR12090432')\n",
      "('SRR12090426', 'SRR12090433')\n",
      "('SRR12090426', 'SRR12090434')\n",
      "('SRR12090426', 'SRR12090435')\n",
      "('SRR12090426', 'SRR18061002')\n",
      "('SRR12090426', 'SRR18061005')\n",
      "('SRR12090427', 'SRR12090428')\n",
      "('SRR12090427', 'SRR12090429')\n",
      "('SRR12090427', 'SRR12090430')\n",
      "('SRR12090427', 'SRR12090431')\n",
      "('SRR12090427', 'SRR12090432')\n",
      "('SRR12090427', 'SRR12090433')\n",
      "('SRR12090427', 'SRR12090434')\n",
      "('SRR12090427', 'SRR12090435')\n",
      "('SRR12090427', 'SRR18061002')\n",
      "('SRR12090427', 'SRR18061005')\n",
      "('SRR12090428', 'SRR12090429')\n",
      "('SRR12090428', 'SRR12090430')\n",
      "('SRR12090428', 'SRR12090431')\n",
      "('SRR12090428', 'SRR12090432')\n",
      "('SRR12090428', 'SRR12090433')\n",
      "('SRR12090428', 'SRR12090434')\n",
      "('SRR12090428', 'SRR12090435')\n",
      "('SRR12090428', 'SRR18061002')\n",
      "('SRR12090428', 'SRR18061005')\n",
      "('SRR12090429', 'SRR12090430')\n",
      "('SRR12090429', 'SRR12090431')\n",
      "('SRR12090429', 'SRR12090432')\n",
      "('SRR12090429', 'SRR12090433')\n",
      "('SRR12090429', 'SRR12090434')\n",
      "('SRR12090429', 'SRR12090435')\n",
      "('SRR12090429', 'SRR18061002')\n",
      "('SRR12090429', 'SRR18061005')\n",
      "('SRR12090430', 'SRR12090431')\n",
      "('SRR12090430', 'SRR12090432')\n",
      "('SRR12090430', 'SRR12090433')\n",
      "('SRR12090430', 'SRR12090434')\n",
      "('SRR12090430', 'SRR12090435')\n",
      "('SRR12090430', 'SRR18061002')\n",
      "('SRR12090430', 'SRR18061005')\n",
      "('SRR12090431', 'SRR12090432')\n",
      "('SRR12090431', 'SRR12090433')\n",
      "('SRR12090431', 'SRR12090434')\n",
      "('SRR12090431', 'SRR12090435')\n",
      "('SRR12090431', 'SRR18061002')\n",
      "('SRR12090431', 'SRR18061005')\n",
      "('SRR12090432', 'SRR12090433')\n",
      "('SRR12090432', 'SRR12090434')\n",
      "('SRR12090432', 'SRR12090435')\n",
      "('SRR12090432', 'SRR18061002')\n",
      "('SRR12090432', 'SRR18061005')\n",
      "('SRR12090433', 'SRR12090434')\n",
      "('SRR12090433', 'SRR12090435')\n",
      "('SRR12090433', 'SRR18061002')\n",
      "('SRR12090433', 'SRR18061005')\n",
      "('SRR12090434', 'SRR12090435')\n",
      "('SRR12090434', 'SRR18061002')\n",
      "('SRR12090434', 'SRR18061005')\n",
      "('SRR12090435', 'SRR18061002')\n",
      "('SRR12090435', 'SRR18061005')\n",
      "('SRR18061002', 'SRR18061005')\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "for ass in assemblies:\n",
    "    runs = list(run_to_ass[run_to_ass['AssemblyID'] == ass]['Run'].values)\n",
    "    # Make pairwise comparisons\n",
    "    pairwise_comps = []\n",
    "    for p in itertools.combinations(runs, 2):\n",
    "        pairwise_comps.append(p)\n",
    "    \n",
    "    for p in pairwise_comps:\n",
    "         print(p)\n",
    "#         os.system(f'differr -a ../runs/{p[0]}.sorted.bam -b ../runs/{p[1]}.sorted.bam -r ../genomes/{ass_to_ref[ass]} -o ../differrs/{p[0]}-{p[1]}_differr.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d2cd13c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "dones = os.listdir('../differrs/')\n",
    "todo_pairwise_comps = []\n",
    "for p in pairwise_comps:\n",
    "    if f'{p[0]}-{p[1]}_differr.txt' not in dones:\n",
    "        todo_pairwise_comps.append(p)\n",
    "        \n",
    "len(todo_pairwise_comps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4a7ab44c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('SRR18061002', 'SRR18061005')]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "todo_pairwise_comps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "602c38b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for p in todo_pairwise_comps:\n",
    "    os.system(f'differr -a ../runs/{p[0]}.sorted.bam -b ../runs/{p[1]}.sorted.bam -r ../genomes/{ass_to_ref[ass]} -o ../differrs/{p[0]}-{p[1]}_differr.txt')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a44392c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR12090426 SRR12090431 8317.402362823486\n",
      "SRR12090426 SRR12090435 8330.892300367355\n",
      "SRR12090426 SRR12090429 8466.323043107986\n",
      "SRR12090426 SRR12090433 8665.718384027481\n",
      "SRR12090427 SRR12090429 9823.660451889038\n",
      "SRR12090427 SRR12090435 10337.222225427628\n",
      "SRR12090427 SRR12090433 10392.66998910904\n",
      "SRR12090426 SRR12090427 10422.965144872665\n",
      "SRR12090427 SRR12090431 10724.460843324661\n",
      "SRR12090426 SRR18061002 6353.305411338806\n",
      "SRR12090426 SRR12090432 8762.288733243942\n",
      "SRR12090426 SRR12090434 8495.181030988693\n",
      "SRR12090426 SRR12090430 9301.804865598679\n",
      "SRR12090427 SRR18061002 7476.335870265961\n",
      "SRR12090426 SRR18061005 18197.32166481018\n",
      "SRR12090427 SRR12090434 9945.002958059311\n",
      "SRR12090427 SRR12090430 10779.975956916809\n",
      "SRR12090426 SRR12090428 10525.690552949905\n",
      "SRR12090427 SRR12090432 10253.605538845062\n",
      "SRR12090428 SRR18061002 7332.118495941162\n",
      "SRR12090429 SRR18061002 5059.6106288433075\n",
      "SRR12090428 SRR12090432 10127.752444505692\n",
      "SRR12090428 SRR12090434 9663.305072784424\n",
      "SRR12090428 SRR12090430 10583.862995624542\n",
      "SRR12090429 SRR12090432 7555.371153831482\n",
      "SRR12090429 SRR12090434 7261.854347229004\n",
      "SRR12090429 SRR12090430 8067.599304676056\n",
      "SRR12090427 SRR12090428 12994.371564626694\n",
      "SRR12090427 SRR18061005 19595.63259100914\n",
      "SRR12090429 SRR12090435 7189.741761922836\n",
      "SRR12090429 SRR12090431 7521.28968834877\n",
      "SRR12090429 SRR12090433 7817.803208112717\n",
      "SRR12090428 SRR12090435 9987.705408334732\n",
      "SRR12090428 SRR12090433 10658.590497732162\n",
      "SRR12090428 SRR12090431 10431.118582963943\n",
      "SRR12090430 SRR12090431 8581.038507938385\n",
      "SRR12090429 SRR18061005 17485.519361019135\n",
      "SRR12090428 SRR12090429 9456.012546300888\n",
      "SRR12090430 SRR12090435 8169.540106058121\n",
      "SRR12090430 SRR12090433 9032.298450946808\n",
      "SRR12090428 SRR18061005 19795.10078048706\n",
      "SRR12090431 SRR12090435 9305.673272132874\n",
      "SRR12090431 SRR12090433 9841.757977962494\n",
      "SRR12090430 SRR12090432 8407.490545988083\n",
      "SRR12090430 SRR18061002 5870.963866949081\n",
      "SRR12090432 SRR18061002 6484.207571268082\n",
      "SRR12090432 SRR12090434 9228.84089255333\n",
      "SRR12090431 SRR18061002 5546.234978914261\n",
      "SRR12090430 SRR12090434 8410.37084531784\n",
      "SRR12090433 SRR12090434 9081.305979967117\n",
      "SRR12090433 SRR18061002 6006.029147863388\n",
      "SRR12090430 SRR18061005 18594.689717769623\n",
      "SRR12090431 SRR12090434 8243.707969427109\n",
      "SRR12090431 SRR18061005 17987.567662477493\n"
     ]
    }
   ],
   "source": [
    "from multiprocessing.dummy import Pool as ThreadPool\n",
    "import time\n",
    "\n",
    "pool = ThreadPool(10)\n",
    "def run_differr(p):\n",
    "    start_time = time.time()\n",
    "    os.system(f'differr -a ../runs/{p[0]}.sorted.bam -b ../runs/{p[1]}.sorted.bam -r ../genomes/{ass_to_ref[ass]} -o ../differrs/{p[0]}-{p[1]}_differr.txt')\n",
    "    print(p[0], p[1], time.time() - start_time)\n",
    "results = pool.map(run_differr, pairwise_comps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a13a0497",
   "metadata": {},
   "source": [
    "SRR12090426 SRR12090431 8317.402362823486\n",
    "SRR12090426 SRR12090435 8330.892300367355\n",
    "SRR12090426 SRR12090429 8466.323043107986\n",
    "SRR12090426 SRR12090433 8665.718384027481\n",
    "SRR12090427 SRR12090429 9823.660451889038\n",
    "SRR12090427 SRR12090435 10337.222225427628\n",
    "SRR12090427 SRR12090433 10392.66998910904\n",
    "SRR12090426 SRR12090427 10422.965144872665\n",
    "SRR12090427 SRR12090431 10724.460843324661\n",
    "SRR12090426 SRR18061002 6353.305411338806\n",
    "SRR12090426 SRR12090432 8762.288733243942\n",
    "SRR12090426 SRR12090434 8495.181030988693\n",
    "SRR12090426 SRR12090430 9301.804865598679\n",
    "SRR12090427 SRR18061002 7476.335870265961\n",
    "SRR12090426 SRR18061005 18197.32166481018\n",
    "SRR12090427 SRR12090434 9945.002958059311\n",
    "SRR12090427 SRR12090430 10779.975956916809\n",
    "SRR12090426 SRR12090428 10525.690552949905\n",
    "SRR12090427 SRR12090432 10253.605538845062\n",
    "SRR12090428 SRR18061002 7332.118495941162\n",
    "SRR12090429 SRR18061002 5059.6106288433075\n",
    "SRR12090428 SRR12090432 10127.752444505692\n",
    "SRR12090428 SRR12090434 9663.305072784424\n",
    "SRR12090428 SRR12090430 10583.862995624542\n",
    "SRR12090429 SRR12090432 7555.371153831482\n",
    "SRR12090429 SRR12090434 7261.854347229004\n",
    "SRR12090429 SRR12090430 8067.599304676056\n",
    "SRR12090427 SRR12090428 12994.371564626694\n",
    "SRR12090427 SRR18061005 19595.63259100914\n",
    "SRR12090429 SRR12090435 7189.741761922836\n",
    "SRR12090429 SRR12090431 7521.28968834877\n",
    "SRR12090429 SRR12090433 7817.803208112717\n",
    "SRR12090428 SRR12090435 9987.705408334732\n",
    "SRR12090428 SRR12090433 10658.590497732162\n",
    "SRR12090428 SRR12090431 10431.118582963943\n",
    "SRR12090430 SRR12090431 8581.038507938385\n",
    "SRR12090429 SRR18061005 17485.519361019135\n",
    "SRR12090428 SRR12090429 9456.012546300888\n",
    "SRR12090430 SRR12090435 8169.540106058121\n",
    "SRR12090430 SRR12090433 9032.298450946808\n",
    "SRR12090428 SRR18061005 19795.10078048706\n",
    "SRR12090431 SRR12090435 9305.673272132874\n",
    "SRR12090431 SRR12090433 9841.757977962494\n",
    "SRR12090430 SRR12090432 8407.490545988083\n",
    "SRR12090430 SRR18061002 5870.963866949081\n",
    "SRR12090432 SRR18061002 6484.207571268082\n",
    "SRR12090432 SRR12090434 9228.84089255333\n",
    "SRR12090431 SRR18061002 5546.234978914261\n",
    "SRR12090430 SRR12090434 8410.37084531784\n",
    "SRR12090433 SRR12090434 9081.305979967117\n",
    "SRR12090433 SRR18061002 6006.029147863388\n",
    "SRR12090430 SRR18061005 18594.689717769623\n",
    "SRR12090431 SRR12090434 8243.707969427109\n",
    "SRR12090431 SRR18061005 17987.567662477493"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
